src/Microsoft.AspNetCore.Authentication.OpenIdConnect/Events/AuthorizationCodeReceivedContext.cs

```diff
@@ -17,29 +18,72 @@ public class AuthorizationCodeReceivedContext : BaseOpenIdConnectContext
         /// <summary>
         /// Creates a <see cref="AuthorizationCodeReceivedContext"/>
         /// </summary>
-        public AuthorizationCodeReceivedContext(HttpContext context, OpenIdConnectOptions options, AuthenticationProperties properties)
+        public AuthorizationCodeReceivedContext(HttpContext context, OpenIdConnectOptions options)
             : base(context, options)
         {
-            Properties = properties;
         }
 
         public AuthenticationProperties Properties { get; set; }
 
         /// <summary>
-        /// Gets or sets the 'code'.
+        /// Gets or sets the <see cref="JwtSecurityToken"/> that was received in the authentication response, if any.
         /// </summary>
-        public string Code { get; set; }
+        public JwtSecurityToken JwtSecurityToken { get; set; }
 
         /// <summary>
-        /// Gets or sets the <see cref="JwtSecurityToken"/> that was received in the id_token + code OpenIdConnectRequest.
+        /// The request that will be sent to the token endpoint and is available for customization.
         /// </summary>
-        public JwtSecurityToken JwtSecurityToken { get; set; }
+        public OpenIdConnectMessage TokenEndpointRequest { get; set; }
+
+        /// <summary>
+        /// The configured communication channel to the identity provider for use when making custom requests to the token endpoint.
+        /// </summary>
+        public HttpClient Backchannel { get; internal set; }
 
         /// <summary>
-        /// Gets or sets the 'redirect_uri'.
+        /// If the developer chooses to redeem the code themselves then they can provide the resulting tokens here. This is the
+        /// same as calling SkipCodeRedemption. If set then the middleware will not attempt to redeem the code. An IdToken
+        /// is required if one had not been previously received in the authorization response. An access token is optional
+        /// if the middleware is to contact the user-info endpoint.
         /// </summary>
-        /// <remarks>This is the redirect_uri that was sent in the id_token + code OpenIdConnectRequest.</remarks>
-        [SuppressMessage("Microsoft.Design", "CA1056:UriPropertiesShouldNotBeStrings", Justification = "user controlled, not necessarily a URI")]
-        public string RedirectUri { get; set; }
+        public OpenIdConnectMessage TokenEndpointResponse { get; set; }
+
+        public bool CodeRedemptionSkipped => TokenEndpointResponse != null;
```

Comment:
It may read better if this is named: SkipCodeRedemption. As in the handler the logic currently reads: if (!CodeRedemptionSkipped) -> redeem code...